### DISCUSSION TITLE
Doubt about implementation of DMP

### DISCUSSION THREAD

**Discussion by ilska** [dfki-ric/movement_primitives#55] [General]
Hi, 

According to the code, the  _dmp.py has been implemented following A.J. Ijspeert, J. Nakanishi, H. Hoffmann, P. Pastor, S. Schaal:
    Dynamical Movement Primitives: Learning Attractor Models for Motor Behaviors (2013), Neural Computation 25(2), pp. 328-373, doi:
    10.1162/NECO_a_00393, https://ieeexplore.ieee.org/document/6797340

I am going through the code and I have doubts about equation 2.3 in the paper. During the calculation of the forcing term, I could not find the (g-y0) term in the imlementation, and I have not been able to find it in the function [dmp_step_euler](https://github.com/dfki-ric/movement_primitives/blob/327b192608747e20e555f59671903854045c4713/movement_primitives/dmp/_dmp.py#L178). I would have expected to have a (g-y0) term in there. 

I was hoping you could help me understand this better. Any assistance you could provide would be greatly appreciated.

**Comment by AlexanderFabisch** [comment]
Hi @ilska ,

you are right. This part is actually similar to [this DMP variation](https://www.heikohoffmann.de/documents/peterdmp.pdf) (Eq. 3). We could document this more clearly or we could introduce a parameter to change it. I am not entirely sure at the moment, but I think this version fixes the scaling issue mentioned in this paper in Fig. 1.

**Comment by ilska** [comment]
Hi @AlexanderFabisch, thanks for your response!

That is what I undestood, that the term K(g-yo) theta of equation 1 in the [paper you link](https://www.heikohoffmann.de/documents/peterdmp.pdf) is added to fix the scaling issues around g = y0. So I tried to find a term related to it when integrating, in the function [dmp_step_euler](https://github.com/dfki-ric/movement_primitives/blob/327b192608747e20e555f59671903854045c4713/movement_primitives/dmp/_dmp.py#L178) and I could not understand why there was nothing, and I thought that I may have missed something :)

**Comment by AlexanderFabisch** [comment]
I'm a bit confused. Eq. 2.3 of Ijspeert et al. is the forcing term (Eq. 3 of Pastor et al.). Eq. 1 of Pastor et al. is the transformation system.

Anyway, I'll check what the effects of the scaling in the forcing term are.

**Comment by AlexanderFabisch** [comment]
Here is what happens when you scale a DMP with $g\approx y_0$ in this implementation:

![Figure_1](https://github.com/dfki-ric/movement_primitives/assets/869592/0cf39b5f-dbac-4286-8934-24185c751129)

And this is the result including the scaling by $(g - y_0)$ in the forcing term:

![Figure_2](https://github.com/dfki-ric/movement_primitives/assets/869592/22322aaf-f595-46fc-a4f6-4d083c6ef6d8)

It's not exactly the same version as in Pastor et al., but it seems to solve the problem.

<details>
<summary>Code</summary>
<code>
import numpy as np
import matplotlib.pyplot as plt
from movement_primitives.dmp import DMP


T = np.linspace(0, 1, 101)
x = np.sin(T ** 2 * 1.99 * np.pi)
y = np.cos(T ** 2 * 1.99 * np.pi)
Y = np.column_stack((x, y))
start = Y[0]
goal = Y[-1]
new_goal = np.array([0.5, 0.5])
Y_shifted = Y - goal[np.newaxis] + new_goal[np.newaxis]

dmp = DMP(n_dims=len(start), execution_time=1.0, dt=0.01, n_weights_per_dim=20)
dmp.imitate(T, Y)
dmp.configure(goal_y=new_goal)
_, Y_dmp = dmp.open_loop()

plt.plot(Y[:, 0], Y[:, 1], label=r"Demonstration, $g \approx y_0$", ls="--")
plt.plot(Y_shifted[:, 0], Y_shifted[:, 1], label="Original shape with new goal", ls="--")
plt.plot(Y_dmp[:, 0], Y_dmp[:, 1], label="DMP with new goal", lw=5, alpha=0.5)
plt.scatter(start[0], start[1], c="r", label="Goal of demonstration: $g$")
plt.scatter(goal[0], goal[1], c="g", label="Start of demonstration: $y_0$")
plt.scatter(new_goal[0], new_goal[1], c="y", label="New goal: $g'$")
plt.xlabel("$y_1$")
plt.ylabel("$y_2$")
plt.legend(loc="best", ncol=2)
plt.xlim((-1.8, 2.1))
plt.ylim((-1.7, 2.2))
plt.tight_layout()
plt.show()
</code>
</details>

**Comment by ilska** [comment]
Ok, thanks! It definitely solves the problem yes. I will try to understand how the current implementation deals with it without adding any term related to it in the transformation system as in Eq 1 in Pastor et al.

Thank you for your time and your super quick answers :)

**Comment by AlexanderFabisch** [comment]
@ilska We are thinking about implementing the term proposed by Pastor et al. in the transformation system. Did you find out more about it? How does it change the scaling?

**Comment by ilska** [comment]
Hi! I'm afraid I haven't had free time to look at it. Most likely, by the end of the month, I'll be able to return back to this. I will write you back with whatever I find :)

**Comment by AlexanderFabisch** [comment]
I was working on this issue for the last days. The term `(goal - start) * phase` is responsible for a smooth scaling between the "old" and "new" trajectory. An implementation is in #33 . You can activate it with `smooth_scaling=True` in each DMP constructor.

**Comment by ilska** [comment]
Great!! I will play with it as soon as I have some time :)
